src/clr-angular/utils/popover/position-operators.spec.ts

```diff
@@ -16,10 +16,57 @@ import {
   flipSidesAndNudgeContent,
   nudgeContent,
   testVisibility,
+  positionsEqual,
+  posiblePositionsAsStrings,
+  transposeStringToPosition,
+  transposePositionToString,
+  SIMPLE_TO_SMART_POSITIONS,
+  validPosition,
 } from './position-operators';
 import { ClrPopoverContentOffset } from './interfaces/popover-content-offset.interface';
 import { ClrViewportViolation } from './enums/viewport-violation.enum';
 
+export function ClrPositionHelperSpec(): void {
+  describe('ClrPosition helper functions', () => {
+    it('should return array of posible positions', () => {
+      expect(posiblePositionsAsStrings()).toEqual(Object.keys(SIMPLE_TO_SMART_POSITIONS));
+    });
+
+    it('should let you find position object by passing string names', () => {
+      posiblePositionsAsStrings().forEach(position => {
+        expect(transposeStringToPosition(position)).toEqual(SIMPLE_TO_SMART_POSITIONS[position]);
+      });
+
+      expect(transposeStringToPosition('NOT_VALID_NAME')).toEqual(null);
+    });
+
+    it('should find string name of position by passing position object', () => {
+      posiblePositionsAsStrings().forEach(position => {
+        expect(position).toBe(transposePositionToString(SIMPLE_TO_SMART_POSITIONS[position]));
+      });
+
+      // Not a valid position
+      expect(null).toBe(transposePositionToString({ axis: -5, side: -5, anchor: -5, content: -5 }));
+    });
+
+    it('should check that the position is valid', () => {
+      expect(validPosition('bottom-right')).toBe(true);
```

Comment:
Maybe we should try all possible positions :(

Reply:
No easy way to check them all without write expectations as a list of keys that we expect to have.

```js
const positionList = Object.keys(SIMPLE_TO_SMART_POSITIONS)
```
`validPosition` just test the key is there. Even could argue that the name is misleading, we don't test is this position a really valid position we just test that the key is there.

```js
Object.keys(SIMPLE_TO_SMART_POSITIONS).forEach((position) => {
  expect(validPosition(position)).toBe(true)
});
```

The problem for me is that this test will always pass :) 

```js
const positionList = ['bottom-left', 'bottom-right' ... ];
```

If we track all keys by hand we always will have to update the list and the test will never fail if we add a new one. It will fail only if we remove one.

```js
expect(positionList.length).toBe(Object.keys(SIMPLE_TO_SMART_POSITIONS).length);
```

This could solve it in a way .. but still, a lot of manual support will be needed to keep this working ...

Reply:
I'm in agreement with Ivan here even though its a lot of manual support. If we are going to provide the adapter to support the existing/legacy positions, imo we should test that 

1) nothing new is added and 
2) everything currently supported still works

This way in a year when someone removes one  or adds a new one (unlikely but you never know) we are aware that something has changed.

Reply:
So we are going with copy the SIMPLE_TO_SMART_POSITIONS into the test and expect to get the same `key => object` as part of the test. 

```js
const TEST_VERSION_OF_SIMPLE_TO_SMART_POSITIONS = { ... }

Object.keys(TEST_VERSION_OF_SIMPLE_TO_SMART_POSITIONS).forEach((positionKey) => {
  expect(SIMPLE_TO_SMART_POSITIONS[positionKey]).toEqual(TEST_VERSION_OF_SIMPLE_TO_SMART_POSITIONS[positionKey]);
})
```
Or could go really crazy 

```js
expect(JSON.stringify(TEST_VERSION_OF_SIMPLE_TO_SMART_POSITIONS)).toBe(JSON.stringify(SIMPLE_TO_SMART_POSITIONS))
```

I just wanted to be clear that I got the idea right!

Reply:
Don't copy, import it directly and use it to test with. 

It should have length of ... 
it should have 'top-left' key && top-left should return ClrPosition { ... } 
etc. 

This way we are directly testing the thing used in the operators and it will be more likely that we catch changes if they happen.